cpmpy/solvers/z3.py

```diff
@@ -368,7 +368,11 @@ def _z3_expr(self, cpm_con, reify=False):
                     return z3.And(self._z3_expr(any(lhs == a for a in rhs.args)),
                                   self._z3_expr(all([lhs <= a for a in rhs.args])))
 
-                lhs, rhs = self._z3_expr(cpm_con.args)
+                st = str(rhs)
+                if str(rhs) == '0' and self.lhs_evaluates_to_bool(cpm_con):
```

Comment:
Also check for other nums instead of only '0' (See main comment in review)
Why not just use `lhs.is_bool()`?

Reply:
lhs.is_bool() is not recursive so it would fail on for example
~(iv[0] + iv[1] == 0)
because is.bool(iv[0] + iv[1] == 0) returns false

Reply:
Not sure what the expected behavior is for Comparion(BoolExpr, num).
for example: Model((iv[1] + iv[0] == 0) == 0) or Model((iv[1] + iv[0] == 0) == 1) work fine
but Model((iv[1] + iv[0] == 0) == 2) also crashes other solvers
should any nonzero integer be considered true here? because then this is a bug when flattening

Reply:
Similarly: Model([(iv[1] + iv[0] == 0) == bv[0], bv[0] == 1]) works if bv is a boolvar but crashes if it is an intvar

Reply:
Hmm, (iv[0] + iv[1]) ==. 0 should definitely return`True` on `is_bool()`
In the Z3 interface we do not apply any flattening to our expressions as their API supports nested expressions by default so we just recursively translate the CPMpy expressions to Z3 expressions.

For the other cases, it might be a bit more tricky indeed. Maybe for now only consider 0 and 1 and raise an error for other values?
I think your last case is related to issue #146, should we fix that bug in this pull request too? Would be most logical to do so I think.